File: cpp/36_有效的数独_medium.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        int rows[9][9];
        int columns[9][9];
        int subBoxs[3][3][9];

        memset(rows, 0, sizeof(rows));
        memset(columns, 0, sizeof(columns));
        memset(subBoxs, 0, sizeof(subBoxs));

        for (int i=0; i<9; i++) {
            for (int j = 0; j<9; j++) {
                auto ch = board[i][j];
                if (ch != '.'){
                    int t = ch - '0' - 1;
                    rows[i][t]++;
                    columns[j][t]++;
                    subBoxs[i/3][j/3][t]++;
                    if (rows[i][t] > 1 || columns[j][t] > 1 || subBoxs[i/3][j/3][t] > 1)
                        return false;
                }
            }
        }
        return true;
    }
};

File: python/36_有效的数独_medium.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        rows, columns = [[0]*9 for i in range(9)], [[0]*9 for i in range(9)]
        subBoxs = [[[0]*9 for i in range(3)] for i in range(3)]

        for i in range(9):
            for j in range(9):
                ch = board[i][j]
                if ch != '.':
                    t = ord(ch) - ord('0') - 1
                    rows[i][t] += 1
                    columns[j][t] += 1
                    subBoxs[i//3][j//3][t] += 1
                    if rows[i][t] > 1 or columns[j][t] > 1 or subBoxs[i//3][j//3][t] > 1:
                        return False
        return True